std/range/package.d

```diff
@@ -1364,183 +1364,189 @@ Params:
+    this(bool r1Chosen, R1 r1, R2 r2)
+    {
+        import std.conv : emplace;
+
+        // This should be the only place r1Chosen is ever assigned
+        // independently
+        this.r1Chosen = r1Chosen;
+        if (r1Chosen)
         {
-            assert(condition);
-            return *cast(R1*) buffer.ptr;
+            _r2 = R2.init;
+            emplace(&_r1, r1);
         }
-        private @property ref R2 r2()
+        else
         {
-            assert(!condition);
-            return *cast(R2*) buffer.ptr;
+            _r1 = R1.init;
+            emplace(&_r2, r2);
         }
+    }
 
-        this(bool condition, R1 r1, R2 r2)
-        {
-            this.condition = condition;
-            import std.conv : emplace;
-            if (condition) emplace(addressOf(this.r1), r1);
-            else emplace(addressOf(this.r2), r2);
-        }
+    // Carefully defined postblit to postblit the appropriate range
+    static if (hasElaborateCopyConstructor!R1
+        || hasElaborateCopyConstructor!R2)
+    this(this)
+    {
+        this.actOnChosen!((ref r) {
+                static if (hasElaborateCopyConstructor!(typeof(r))) r.__postblit();
+            });
+    }
+
+    static if (hasElaborateDestructor!R1 || hasElaborateDestructor!R2)
+    ~this()
+    {
+        actOnChosen!((ref r) => destroy(r))(this);
+    }
 
-        // Carefully defined postblit to postblit the appropriate range
-        static if (hasElaborateCopyConstructor!R1
-            || hasElaborateCopyConstructor!R2)
-        this(this)
+    static if (isInfinite!R1 && isInfinite!R2)
+        // Propagate infiniteness.
+        enum bool empty = false;
+    else
+        @property bool empty()
         {
-            if (condition)
-            {
-                static if (hasElaborateCopyConstructor!R1) r1.__postblit();
-            }
-            else
-            {
-                static if (hasElaborateCopyConstructor!R2) r2.__postblit();
-            }
+            return actOnChosen!(r => r.empty)(this);
         }
 
-        static if (hasElaborateDestructor!R1 || hasElaborateDestructor!R2)
-        ~this()
+    @property auto ref front()
+    {
+        static auto ref getFront(R)(ref R r) { return r.front; }
+        return actOnChosen!getFront(this);
+    }
+
+    void popFront()
+    {
+        return actOnChosen!((ref r) { r.popFront; })(this);
+    }
+
+    static if (isForwardRange!R1 && isForwardRange!R2)
+        @property auto save()
         {
-            if (condition) destroy(r1);
-            else destroy(r2);
+            auto result = this;
+            actOnChosen!((ref r) { r = r.save; })(result);
+            return result;
         }
 
-        static if (isInfinite!R1 && isInfinite!R2)
-            // Propagate infiniteness.
-            enum bool empty = false;
-        else
-            @property bool empty()
-            {
-                return condition ? r1.empty : r2.empty;
-            }
+    @property void front(T)(T v)
```

Comment:
no need for this if the other `front` returns ref

Reply:
Setters take precedence over `ref` getters, so in the case where R1 defines a setter and a `ref` getter we should call the setter. If we had just the `ref` getter here then R1's setter would never be called. In the case where R1 defines *only* a `ref` getter, this overload does nothing. In the case where R1 defines *only* a setter, this overload is the only route to it. By symmetry the same holds for (R1, R2) x (front, back).

I guess I could do something that detects if there's actually a setter for either of R1 or R2? I'm not sure what's the best way to do that.